### PR TITLE
Stacking machine rework

### PR DESCRIPTION
**Pull request by DerelictDrone**
While the panel is open, you can:
Use your hand to turn the input/output.
Use a multitool to change between turning input or output.
Use a wrench to change stack_amt.
And of course crowbar it to deconstruct!

This might not be the best possible way of doing it, but it's the only way I know how, if there's anything that could be done better(like the descriptions) feel free to suggest.

Sprites as seen in-game
image
Full description with steps provided:
image

:cl:  
rscadd: You can now deconstruct and reconstruct stacking machines, as well as change their stack size, and the input or output directions separately
/:cl:

- [x] Construction and deconstruction with sprites
- [x] Setting direction and stack size
- [x] The ability to link Stackers to a silo.

### CODE REVIEW

**Comment by MegaEmpirical** [comment]
stack machine needs more love, thanks

**Comment by DerelictDrone** [comment]
Maybe once I learn a little bit more I can make a toggle with wirecutters so it can stack /stack/ items as well as sheets, so it can stack ore, marker beacons, and cable coil.

Main issue there is that the stacky boi drops one stack no matter what.

If I come back and touch this again it'll make it so it drops stack_amt of items rather than one item with the amount variable changed to stack_amt. 

So it can do more than one stack at a time, and also not eat items when it's set to a stack_amt higher than the max amount a stack can hold 
(50 cable coil, one stack of 30 one stack of 20 enter a stacking machine set to stack 50. Stack machine outputs 30 cable coil and deletes the extra 20, that is the problem with dropping one item with a set amount)

**Comment by boodaliboo** [comment]
good first pr

**Comment by DerelictDrone** [comment]
> 
> 
> good first pr

https://github.com/yogstation13/Yogstation/pull/7077 was my first PR, this is my second.

**Comment by DerelictDrone** [comment]
Gonna re-open this when I have the two features mentioned above written, working, and tested.

Edit: Nevermind I'll save this for later I'm not good enough yet.

**Comment by alexkar598** [comment]
This Pull Request has been marked as stale due to lack of commit activity, if no further work is done within a week, this PR will be closed at which point you may contact a maintainer on discord to have this PR reopened or simply make a new one.

**Comment by DerelictDrone** [comment]
Sorry about the wait, I've been dragged around a lot the past week or two.
Unless anything needs to be corrected, that's the last commit I need to make.

File: code/modules/mining/machine_stacking.dm [hunk 1]
```diff
@@ -81,24 +86,34 @@
 	var/stack_amt = 50 //amount to stack before releassing
 	var/datum/component/remote_materials/materials
 	var/force_connect = FALSE
+	var/io = 0 //This is used for determining whether we change the Input or Output
+
+
+/obj/machinery/mineral/stacking_machine/examine(mob/user)
+	. = ..()
+	if(stack_amt != 1)
+		. += {"<span class='notice'>Currently outputting stacks at <b>[stack_amt] sheets</b><span>"}
```
Comment (alexkar598) [inline code/modules/mining/machine_stacking.dm]: Replace this if/else block with this

```suggestion
		. += {"<span class='notice'>Currently outputting stacks at <b>[stack_amt] sheet[(stack_amt > 1) ? "s" : ""]</b><span>"}
```

File: code/modules/mining/machine_stacking.dm [hunk 2]
```diff
@@ -81,24 +86,34 @@
...
+The I/O setting can be <b>changed with a multitool</b> to change what the dial controls
+There are some <b>bolts</b> to limit stack size."}
 
 /obj/machinery/mineral/stacking_machine/Initialize(mapload)
 	. = ..()
 	proximity_monitor = new(src, 1)
 	materials = AddComponent(/datum/component/remote_materials, "stacking", mapload, FALSE, mapload && force_connect)
 
 /obj/machinery/mineral/stacking_machine/HasProximity(atom/movable/AM)
+	if(panel_open || !powered())
```
Comment (alexkar598) [inline code/modules/mining/machine_stacking.dm]:
```suggestion
	if(panel_open || !(stat & (BROKEN|NOPOWER)))
```

File: code/modules/mining/machine_stacking.dm [hunk 3]
```diff
@@ -120,3 +135,52 @@
...
+		return
+	if(default_deconstruction_crowbar(W))
+		return
+
+	if(!powered())
+		return
+
+	if(W.tool_behaviour == TOOL_WRENCH && panel_open)
+		var/stsize = input(user, "How much should [src] stack to? (1-50)", "Stack size") as null|num
+		if(stsize && (stsize > 0 && stsize <= 50))
```
Comment (alexkar598) [inline code/modules/mining/machine_stacking.dm]: Move this check further down

```suggestion
		if(stsize)
```

File: code/modules/mining/machine_stacking.dm [hunk 4]
```diff
@@ -120,3 +135,52 @@
...
+			stack_amt = stsize
```
Comment (alexkar598) [inline code/modules/mining/machine_stacking.dm]:
```suggestion
			stack_amt = CLAMP(stsize,1,50)
```

File: code/modules/mining/machine_stacking.dm [hunk 5]
```diff
@@ -120,3 +135,52 @@
...
+			if(stack_amt != 1)
```
Comment (alexkar598) [inline code/modules/mining/machine_stacking.dm]: Replace this if/else block with this
```suggestion
			to_chat(user, "<span class='notice'>[src] is now set to output <b>[stack_amt] sheet[(stack_amt > 1) ? "s" : ""]</b><span>")
```

File: code/modules/mining/machine_stacking.dm [hunk 6]
```diff
@@ -120,3 +135,52 @@
...
+				to_chat(user, "<span class='notice'>[src] is now set to output <b>[stack_amt] sheets</b><spanclass>")
+			else if(stack_amt == 1)
+				to_chat(user, "<span class='notice'>[src] is now set to output <b>[stack_amt] sheet</b></span>")
+			return
+
+	return ..()
+
+/obj/machinery/mineral/stacking_machine/attack_hand(mob/user)
```
Comment (alexkar598) [inline code/modules/mining/machine_stacking.dm:158]:
```suggestion
/obj/machinery/mineral/stacking_machine/attack_hand(mob/user)
	if(!panel_open)
		return ..()
```

File: code/modules/mining/machine_stacking.dm [hunk 7]
```diff
@@ -120,3 +135,52 @@
...
+	if(panel_open &&(io == 0))
```
Comment (alexkar598) [inline code/modules/mining/machine_stacking.dm]:
```suggestion
	if(o == 0)
```

File: code/modules/mining/machine_stacking.dm [hunk 8]
```diff
@@ -81,24 +86,34 @@
 	var/stack_amt = 50 //amount to stack before releassing
 	var/datum/component/remote_materials/materials
 	var/force_connect = FALSE
+	var/io = 0 //This is used for determining whether we change the Input or Output
```
Comment (alexkar598) [inline code/modules/mining/machine_stacking.dm]: Replace your magic numbering of 0 and 1 with a DEFINE

File: code/modules/mining/machine_stacking.dm [hunk 9]
```diff
@@ -120,3 +135,52 @@
...
+	return ..()
+
+/obj/machinery/mineral/stacking_machine/attack_hand(mob/user)
+	if(panel_open &&(io == 0))
+		input_dir = turn(input_dir, -90)
+		if(input_dir == output_dir) //Input and output can't be the same or you create the immovable sheet.
+			input_dir = turn(input_dir, -90)
+		to_chat(user, "<span class='notice'>You set [src]'s input to take from the [dir2text(input_dir)].</span>")
+		return
+	else if (panel_open && (io == 1))
```
Comment (alexkar598) [inline code/modules/mining/machine_stacking.dm]:
```suggestion
	else if (io == 1)
```

**Review by alexkar598 — CHANGES_REQUESTED** [review]
Il do some more checking on the return values later but fix these in the time being

File: code/modules/mining/machine_stacking.dm [hunk 10]
```diff
@@ -81,24 +88,31 @@
 	var/stack_amt = 50 //amount to stack before releassing
 	var/datum/component/remote_materials/materials
 	var/force_connect = FALSE
+	var/io = INPUT //This is used for determining whether we change the Input or Output
+
+
+/obj/machinery/mineral/stacking_machine/examine(mob/user)
+	. = ..()
+		. += {"<span class='notice'>Currently outputting stacks at <b>[stack_amt] sheet[(stack_amt > 1) ? "s" : ""]</b><span>"}
```
Comment (alexkar598) [inline code/modules/mining/machine_stacking.dm]:
```suggestion
		. += "<span class='notice'>Currently outputting stacks at <b>[stack_amt] sheet[(stack_amt > 1) ? "s" : ""]</b><span>"
```

File: code/modules/mining/machine_stacking.dm [hunk 11]
```diff
@@ -81,24 +88,31 @@
...
+	if(panel_open)
+		. += {"The I/O is set to change [io ? "output" : "input"] currently.
```
Comment (alexkar598) [inline code/modules/mining/machine_stacking.dm]:
```suggestion
		. += "The I/O is set to change [io ? "output" : "input"] currently. \
```

File: code/modules/mining/machine_stacking.dm [hunk 12]
```diff
@@ -81,24 +88,31 @@
...
+Input is <b>[dir2text(input_dir)]</b>
```
Comment (alexkar598) [inline code/modules/mining/machine_stacking.dm]:
```suggestion
Input is <b>[dir2text(input_dir)]</b> \
```

File: code/modules/mining/machine_stacking.dm [hunk 13]
```diff
@@ -81,24 +88,31 @@
...
+Output is <b>[dir2text(output_dir)]</b>
```
Comment (alexkar598) [inline code/modules/mining/machine_stacking.dm]:
```suggestion
Output is <b>[dir2text(output_dir)]</b> \
```

File: code/modules/mining/machine_stacking.dm [hunk 14]
```diff
@@ -81,24 +88,31 @@
...
+There is a dial that can be <b>turned by hand</b> to change direction.
```
Comment (alexkar598) [inline code/modules/mining/machine_stacking.dm]:
```suggestion
There is a dial that can be <b>turned by hand</b> to change direction. \
```

File: code/modules/mining/machine_stacking.dm [hunk 15]
```diff
@@ -81,24 +88,31 @@
...
+The I/O setting can be <b>changed with a multitool</b> to change what the dial controls
```
Comment (alexkar598) [inline code/modules/mining/machine_stacking.dm]:
```suggestion
The I/O setting can be <b>changed with a multitool</b> to change what the dial controls \
```

File: code/modules/mining/machine_stacking.dm [hunk 16]
```diff
@@ -81,24 +88,31 @@
...
+There are some <b>bolts</b> to limit stack size."}
```
Comment (alexkar598) [inline code/modules/mining/machine_stacking.dm]:
```suggestion
There are some <b>bolts</b> to limit stack size."
```

**Review by alexkar598 — CHANGES_REQUESTED** [review]
thats not how multiline strings work

File: code/modules/mining/machine_stacking.dm (hunk 16 above)
Comment (DerelictDrone) [reply code/modules/mining/machine_stacking.dm]: If I apply all of these together it just looks like a word salad to be honest.
image
But if this is how to keep it consistent with other machines, that would probably be best.

Comment (alexkar598) [reply code/modules/mining/machine_stacking.dm]: In that case, for each line do . += "message here"

Comment (DerelictDrone) [reply code/modules/mining/machine_stacking.dm]: image
This is what it looks like now.
Let me know if this needs to be changed at all.

File: code/modules/mining/machine_stacking.dm [hunk 17]
```diff
@@ -81,24 +88,31 @@
 	var/stack_amt = 50 //amount to stack before releassing
 	var/datum/component/remote_materials/materials
 	var/force_connect = FALSE
+	var/io = INPUT //This is used for determining whether we change the Input or Output
+
+
+/obj/machinery/mineral/stacking_machine/examine(mob/user)
+	. = ..()
+		. += "<span class='notice'>Currently outputting stacks at <b>[stack_amt] sheet[(stack_amt > 1) ? "s" : ""]</b><span>"
```
Comment (alexkar598) [inline code/modules/mining/machine_stacking.dm]:
```suggestion
	. += "<span class='notice'>Currently outputting stacks at <b>[stack_amt] sheet[(stack_amt > 1) ? "s" : ""]</b><span>"
```

File: code/modules/mining/machine_stacking.dm [hunk 18]
```diff
@@ -61,9 +61,16 @@
 	src.updateUsrDialog()
 	return
 
+/obj/machinery/mineral/stacking_unit_console/attackby(obj/item/W, mob/user, params)
+	if(default_deconstruction_screwdriver(user, "console-open", "console", W))
+		updateUsrDialog()
+		return
+	if(default_deconstruction_crowbar(W))
+		return
```
Comment (alexkar598) [inline code/modules/mining/machine_stacking.dm:69]:
```suggestion
		return
	return ..()
```

File: code/modules/mining/machine_stacking.dm [hunk 19]
```diff
@@ -120,3 +134,51 @@
...
+			input_dir = turn(input_dir, -90)
+		to_chat(user, "<span class='notice'>You set [src]'s input to take from the [dir2text(input_dir)].</span>")
+		return
+	else if (io == OUTPUT)
+		output_dir = turn(output_dir, -90)
+		if(input_dir == output_dir)
+			output_dir = turn(output_dir, -90)
+		to_chat(user, "<span class='notice'>You set [src]'s output to drop stacks [dir2text(output_dir)].</span>")
+		return
+	return TRUE
```
Comment (alexkar598) [inline code/modules/mining/machine_stacking.dm]:
```suggestion
	return ..()
```

**Review by alexkar598 — CHANGES_REQUESTED** [review]
Should be the last few changes

**Pull request merged**